### PR TITLE
Docs on how to install Intel's SYCL with POCL on ARM

### PR DESCRIPTION
**Pull request by jandrovins**
Added documentation on how to use POCL as SYCL's OpenCL backend on ARM, as kindly requested by @pjaaskel. Let me know what you think, and if you want any changes.

### CODE REVIEW

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+   export PKG_CONFIG_PATH=${BASE_PATH}/lib64/pkgconfig/:${BASE_PATH}/lib/pkgconfig/:${BASE_PATH}/share/pkgconfig/:$PKG_CONFIG_PATH
+
+   export CC=clang CXX=clang++
+
+Installation of vanilla LLVM
+----------------------------
+
+We will install LLVM 16 (vanilla version, not Intel's). This is the LLVM that POCL will use,
+and is needed to compile POCL. Notice that we want a static LLVM.
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:131]:
```suggestion
and is needed to compile PoCL. Notice that we want a static LLVM.
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+   cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=OFF -DLLVM_OPTIMIZED_TABLEGEN=ON \
+   -DLLVM_TARGETS_TO_BUILD=AArch64 -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra" -DLLVM_BUILD_TOOLS=ON \
+   -DLLVM_ENABLE_RUNTIMES="openmp" -DBUILD_SHARED_LIBS=OFF -DLLVM_ENABLE_EH=ON -DLLVM_ENABLE_RTTI=ON \
+   -DLLVM_PARALLEL_LINK_JOBS=48 -DCMAKE_INSTALL_PREFIX=<YOUR VANILLA LLVM PREFIX> -DLLVM_ENABLE_DOXYGEN=OFF \
+   -DLLVM_ENABLE_SPHINX=OFF -DLLVM_ENABLE_LLD=OFF -DLLVM_ENABLE_BINDINGS=OFF -DLLVM_ENABLE_LIBXML2=OFF \
+   -DOPENMP_ENABLE_LIBOMPTARGET=OFF -DLLVM_STATIC_LINK_CXX_STDLIB=ON ../llvm
+
+3. ninja install
+
+Installation of POCL
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:150]:
```suggestion
Installation of PoCL
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+   `OpenCL-ICD-Loader <https://github.com/KhronosGroup/OpenCL-ICD-Loader>`__
+   and
+   `OpenCL-Headers <https://github.com/KhronosGroup/OpenCL-Headers>`__.
+   Installation is simple, but you should have a specific git checkout
+   for both repositories depending on your Intel LLVM version. The git
+   checkouts can be found at the `Intel LLVM repository, at file
+   ./opencl/CMakeLists.txt, at lines 23 and
+   24 <https://github.com/intel/llvm/blob/c5d04bcc0b7b7adf93a9f4c57faf6fada06575cf/opencl/CMakeLists.txt#L23>`__.
+
+POCL installation
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:170]:
```suggestion
PoCL installation
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+   VVV_ICD_LOADER=<YOUR ICD-LOADER PREFIX>
+   VVV_OCL_HEADERS=<YOUR OPENCL HEADERS PREFIX>
+   export CPLUS_INCLUDE_PATH=${VVV_OCL_HEADERS}/include:$CPLUS_INCLUDE_PATH/
+   export C_INCLUDE_PATH=${VVV_OCL_HEADERS}/include:$CPLUS_INCLUDE_PATH/
+   export PKG_CONFIG_PATH=${VVV_ICD_LOADER}/lib64/pkgconfig/:${VVV_OCL_HEADERS}/share/pkgconfig:$PKG_CONFIG_PATH
+   export LIBRARY_PATH=${VVV_ICD_LOADER}/lib64/:$LIBRARY_PATH
+   export LD_LIBRARY_PATH=${VVV_ICD_LOADER}/lib64:$LD_LIBRARY_PATH
+
+2. Clone the `POCL repository <https://github.com/pocl/pocl>`__, create
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:186]:
```suggestion
2. Clone the `PoCL repository <https://github.com/pocl/pocl>`__, create
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+::
+
+   CC="clang" CXX="clang++" cmake -G Ninja -DCMAKE_INSTALL_PREFIX=<YOUR POCL INSTALLATION PREFIX> \
+   -DENABLE_ICD=ON -DCMAKE_PREFIX_PATH="<YOUR VANILLA LLVM PREFIX>;$VVV_ICD_LOADER;$VVV_OCL_HEADERS" \
+   -DENABLE_SPIR=ON -DENABLE_SPIRV=ON -DLLVM_SPIRV=<FULL PATH TO THE llvm-spirv BINARY FROM INTEL LLVM>  \
+   -DCMAKE_BUILD_TYPE=RelWithDebInfo -DSTATIC_LLVM=ON ..
+
+3. ninja install
+4. OPTIONAL: Run the test suite: ``ctest -j 128 -L internal``.
+5. When using POCL you should export some environment variables. I
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:198]:
```suggestion
5. When using PoCL you should export some environment variables. I
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+   export PKG_CONFIG_PATH=${BASE_PATH}/lib64/pkgconfig/::$PKG_CONFIG_PATH
+
+   export OCL_ICD_VENDORS=${BASE_PATH}/etc/OpenCL/vendors/
+
+   # Variables for debugging programs
+   export VVV_pocl_help="SYCL_PI_TRACE=2 POCL_DEBUG=all OCL_ICD_ENABLE_TRACE=1"
+
+
+Using SYCL with POCL
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:227]:
```suggestion
Using SYCL with PoCL
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+--------------------
+
+I uploaded a simple example to test if SYCL is working with POCL. It
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:230]:
```suggestion
I uploaded a simple example to test if SYCL is working with PoCL. It
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+just tests that you can compile and execute simple SYCL programs using
+POCL as the OpenCL implementation that SYCL uses.
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:232]:
```suggestion
PoCL as the OpenCL implementation that SYCL uses.
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+1. ``source env-pocl.sh``
+2. ``source env-sycl.sh``
+3. The first test you should do is validate that the SYCL runtime can
+   find and query simple information from the POCL runtime. This is how
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:237]:
```suggestion
   find and query simple information from the PoCL runtime. This is how
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+      return 0;
+    }
+
+
+5. ``clang++ -fsycl t.cpp && ./a.out``
+
+Known issues
+------------
+
+1. ``queue.memset()`` is not supported using POCL right now (see
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:299]:
```suggestion
1. ``queue.memset()`` is not supported using PoCL right now (see 
```

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+1. ``source env-pocl.sh``
+2. ``source env-sycl.sh``
+3. The first test you should do is validate that the SYCL runtime can
+   find and query simple information from the POCL runtime. This is how
+   it looks for me:
+
+::
+
+   [host@user]$ sycl-ls
+   [opencl:cpu:0] Portable Computing Language, pthread-0xd01 OpenCL 3.0 PoCL HSTR: pthread-aarch64-unknown-linux-gnu-tsv110 [4.0-pre next-0-gbbb3d72]
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:243]: The PoCL version is different here which was listed above. Perhaps you can just more vaguely state it was tested with 4.0 as we are very close to branching?

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+
+4. After that, you can try with this test:
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:245]: Can you in addition write something about the DPC++ SYCL tests that are expected to pass with ARM?

File: doc/sphinx/source/sycl_with_pocl_arm.rst [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+See `issue #9099 on Intel's LLVM repository <https://github.com/intel/llvm/issues/9099>`__.
+
+3. When trying to compile with ``-O0`` you will get a runtime error 
+saying that some OpenCL kernel has an ``undefined symbol: _group_id_x``.
+
+Troubleshooting
+---------------
+
+If you ever have runtime errors, these errors can be located at three
+parts: they might be in the ICD-Loader, in POCL or in SYCL.
```
Comment (pjaaskel) [inline doc/sphinx/source/sycl_with_pocl_arm.rst:316]:
```suggestion
parts: they might be in the ICD-Loader, in PoCL or in SYCL.
```

**Review by pjaaskel — CHANGES_REQUESTED** [review]
This is excellent, thanks! Mostly nitpicks about POCL -> PoCL. Can you also add a sentence to CHANGES about fixing issues with SYCL via DPC++ on PoCL on ARM? Would be nice to advertise to other ARM users.

File: doc/sphinx/source/sycl_with_pocl_arm.rst (hunk 12 above)
Comment (jandrovins) [reply doc/sphinx/source/sycl_with_pocl_arm.rst:245]: You mean these ones? https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL

Comment (pjaaskel) [reply doc/sphinx/source/sycl_with_pocl_arm.rst:245]: Right.

**Comment by franz** [comment]
Thanks @jandrovins, i have rebased this onto release_4_0, added some notes on the integrated SYCL testsuites and merged it.

**Pull request closed without merging**